### PR TITLE
Refactor Page Allocator

### PR DESCRIPTION
**Pull request by Tangzh33**
This pull request is associated with issue #1044. Here is the detailed progress:

- [x]  Safe Implementation of buddy system
    - [x]  Confirmation: Current `FrameAllocator` manage pages with `BTreeSet` instead of `linked list`. i.e., current `FrameAllocator` from `buddy_system_allocator` crate is safe.
- [x]  Redesign the Page Allocator Interface
    - [x]  Design:
        - [x]  Suitable abstraction for the behavior of the page allocator
    - [x]  Implementation:
        - [x]  Ensure compatibility between buddy frame allocator and the new PageAlloc Trait
        - [x]  Utilize the original `buddy_system_allocator` crate as bootstrap policies in `ostd`
        - [x]  Decouple bootstrap frame allocator from formal frame allocators
        - [x]  Move current `FrameAllocator` out of `ostd`
- [x]  Resolve circular dependencies in Bootstrap Allocator
    - [x]  Design a new bootstrap page allocator that operates independently of the heap allocator(may be static bitmap)
    - [x]  Implementation and Debugging

### CODE REVIEW

**Comment by tatetian** [comment]
@Tangzh33 As I understand it, this PR isn't ready for reviewing. For one thing, it contains a merge commit. And the commit messages do not follow the current convention.

**Comment by Tangzh33** [comment]
> @Tangzh33 As I understand it, this PR isn't ready for reviewing. For one thing, it contains a merge commit. And the commit messages do not follow the current convention. So could you mark it as a _draft_ PR?

Get it. I will resquash, recommit and rebase the commits until the work is ready for review.

**Comment by Tangzh33** [comment]
I have meticulously restructured the commits to align with the community's standards. Presently, this pull request (PR) has encountered failures in the CI tests, primarily due to two issues:

1. The failure in the `ktest CI` suite is attributed to the decoupling of `ostd` from the page allocator's implementation in this PR. As a result, the current tests are failing due to the absence of the `__ostd_page_allocator_init` function during the linking phase. It is imperative to manually integrate the relevant crate with the foundational crate when executing unit tests. After dicussions with @junyang-zh, we have resolved to refactor the ktest process (PR#1159). Following the merge of the aforementioned PR, I will proceed to address the pertinent bugs.
2. The failure in the `ostd-publish CI` is due to the introduction of a new procedural macro, `ostd_macros::page_allocator_init`, which has yet to be published in the utilities. This issue can be promptly fixed once the review process is complete.

Based on the analysis, the problems we are currently confronting either depend on to-be-merged code, or can be simply fixed after fully discussion. I suppose that this PR has already prepared for review.

File: ostd/src/mm/page/meta.rs [hunk 1]
```diff
@@ -295,9 +295,8 @@ fn alloc_meta_pages(nframes: usize) -> Vec<Paddr> {
         .get()
         .unwrap()
         .lock()
-        .alloc(nframes)
-        .unwrap()
-        * PAGE_SIZE;
+        .alloc(core::alloc::Layout::from_size_align(nframes * PAGE_SIZE, PAGE_SIZE).unwrap())
+        .unwrap_or_else(|| panic!("Failed to allocate metadata pages"));
```
Comment (junyang-zh) [inline ostd/src/mm/page/meta.rs]: Would be the same as `.expect("Failed to allocate metadata pages")`. Didn't Clippy instructed you to do so?

File: ostd/src/mm/page_table/node.rs [hunk 2]
```diff
@@ -231,8 +231,18 @@ where
     /// set the lock bit for performance as it is exclusive and unlocking is an
     /// extra unnecessary expensive operation.
     pub(super) fn alloc(level: PagingLevel) -> Self {
-        let meta = PageTablePageMeta::new_locked(level);
-        let page = page::allocator::alloc_single::<PageTablePageMeta<E, C>>(meta).unwrap();
+        let page = page::allocator::PAGE_ALLOCATOR.get().unwrap().lock().alloc_page(PAGE_SIZE).map(|paddr|{
+            Page::<PageTablePageMeta<E, C>>::from_unused(paddr)
```
Comment (junyang-zh) [inline ostd/src/mm/page_table/node.rs]: Firstly, the `page::allocator::alloc_single::<PageTablePageMeta<E, C>>(meta)` should be kept for allocating typed pages easier. Secondly, where's the `from_unused` function that receives initial metadata?

File: ostd/src/mm/page_table/boot_pt.rs [hunk 3]
```diff
@@ -151,22 +160,39 @@ impl<E: PageTableEntryTrait, C: PagingConstsTrait> BootPageTable<E, C> {
...
         frame
     }
-}
 
-impl<E: PageTableEntryTrait, C: PagingConstsTrait> Drop for BootPageTable<E, C> {
-    fn drop(&mut self) {
-        for frame in &self.frames {
-            PAGE_ALLOCATOR.get().unwrap().lock().dealloc((*frame) * PAGE_SIZE, 1);
-        }
+    pub fn from_paddr_to_meta(&mut self) {
```
Comment (junyang-zh) [inline ostd/src/mm/page_table/boot_pt.rs]: I think the behavior should be documented.

File: ostd/src/mm/page_table/boot_pt.rs [hunk 4]
```diff
@@ -151,22 +160,39 @@ impl<E: PageTableEntryTrait, C: PagingConstsTrait> BootPageTable<E, C> {
...
+//         for frame in &self.frames {
+//             PAGE_ALLOCATOR
+//                 .get()
+//                 .unwrap()
+//                 .lock()
+//                 .dealloc((*frame) * PAGE_SIZE, 1);
+//         }
+//     }
+// }
+
```
Comment (junyang-zh) [inline ostd/src/mm/page_table/boot_pt.rs]: Remove it.

File: ostd/src/mm/page/mod.rs [hunk 5]
```diff
@@ -83,6 +83,30 @@ impl<M: PageMeta> Page<M> {
...
 
+    /// Check whether the page described by paddr is allocated.
+    /// true if the page is allocated, false otherwise.
+    /// 
+    /// # Panics
+    /// 
+    /// The function panics if:
+    /// 
+    /// - the physical address is out of bound or not aligned.
+    pub fn check_page_status(paddr: Paddr) -> bool {
```
Comment (junyang-zh) [inline ostd/src/mm/page/mod.rs:94]: A way to ensure that the status of the page will not be changed after the load would be marking this function unsafe. The caller should ensure that no other threads will do `from_unused` or `drop`. So it is better to confine the `from_unused` calls into a single module with a given set of APIs, like in the old way of `page::allocator`, with a few exceptions. The caller should take a big lock on the global page allocator too.

File: kernel/libs/page_allocator/src/buddy_allocator.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            self.free_list[size.trailing_zeros() as usize].insert(current_start);
+            current_start += size;
+        }
+
+        self.total += total;
+    }
+
+    /// Allocate a range of frames from the allocator, returning the first frame of the allocated
+    /// range.
+    pub fn alloc(&mut self, count: usize) -> Option<usize> {
```
Comment (junyang-zh) [inline kernel/libs/page_allocator/src/buddy_allocator.rs:69]: You should differentiate `count` and `size`, the former should mean the number of pages to be allocated (which is not quite good), and the latter should mean the total number of bytes of the pages to be allocated.

File: ostd/src/lib.rs [hunk 7]
```diff
@@ -79,9 +79,6 @@ pub fn init() {
     let meta_vec = mm::init_page_meta();
     mm::page::allocator::init();
     mm::kspace::init_kernel_page_table(meta_vec);
-    // SAFETY: no CPU local objects have been accessed by this far. And
-    // we are on the BSP.
-    unsafe { cpu::cpu_local::init_on_bsp() };
```
Comment (junyang-zh) [inline ostd/src/lib.rs]: So where is it now?

File: ostd/src/mm/page_table/node.rs [hunk 8]
```diff
@@ -231,18 +231,18 @@ where
...
+            .unwrap()
+            .lock()
+            .alloc_page(PAGE_SIZE)
+            .map(|paddr| {
+                Page::<PageTablePageMeta<E, C>>::from_unused(
+                    paddr,
+                    PageTablePageMeta::new_locked(level),
+                )
+            })
+            .unwrap_or_else(|| panic!("Failed to allocate a page for a page table node"));
```
Comment (junyang-zh) [inline ostd/src/mm/page_table/node.rs:245]: Don't do `.unwrap_or_else(|| panic!(`. Also, `from_unused` shouldn't be scattered all over the place.

File: ostd/src/lib.rs (hunk 7 above)
Comment (Tangzh33) [reply ostd/src/lib.rs]: It is mistakenly added while rebasing, which is originated from former implementation and deprecated in the latest.

File: ostd/src/mm/frame/options.rs [hunk 9]
```diff
@@ -56,13 +60,29 @@ impl FrameAllocOptions {
...
+            PAGE_ALLOCATOR
+                .get()
+                .unwrap()
+                .lock()
+                .alloc(Layout::from_size_align(size, PAGE_SIZE).unwrap())
+                .map(|begin_paddr| {
+                    cont_pages::ContPages::from_unused(begin_paddr..begin_paddr + size, |_| {
+                        FrameMeta::default()
+                    })
+                })
```
Comment (junyang-zh) [inline ostd/src/mm/frame/options.rs:74]: You should have implemented the `ostd::page::allocator::alloc(layout, meta_fn)` API, you should use that here.

I think it still necessary to have a set of `ostd::page::allocator::alloc_` API that wraps over the `from_unused` and the `PAGE_ALLOCATOR` call. As for the boot page allocator, accessing `BOOT_PAGE_ALLOCATOR` directly is fine.

File: ostd/src/mm/page/allocator.rs [hunk 10]
```diff
@@ -333,6 +343,10 @@ pub(crate) struct BootstrapFrameAllocator {
     frame_cursor: usize,
 }
 
+#[export_name = "BOOTSTRAP_PAGE_ALLOCATOR"]
```
Comment (junyang-zh) [inline ostd/src/mm/page/allocator.rs]: I think that the symbol shouldn't be exported.

File: ostd/src/boot/smp.rs [hunk 11]
```diff
@@ -61,15 +64,27 @@ pub fn boot_all_aps() {
...
-        let boot_stack_array =
-            page::allocator::alloc_contiguous(2 * PAGE_SIZE, |_| KernelMeta::default()).unwrap();
+        let boot_stack_array = allocator
+            .alloc(Layout::from_size_align(2 * PAGE_SIZE, PAGE_SIZE).unwrap())
+            .map(|paddr| {
+                page::ContPages::from_unused(paddr..paddr + 2 * PAGE_SIZE, |_| {
+                    KernelMeta::default()
+                })
+            })
+            .unwrap();
```
Comment (junyang-zh) [inline ostd/src/boot/smp.rs:76]: Here it should be able to use the `page::allocator::alloc_contiguous` API.

File: ostd/src/mm/page/allocator.rs [hunk 12]
```diff
@@ -5,128 +5,236 @@
...
-        })
+        .alloc_page(align)
+        .map(|paddr| Page::from_unused(paddr, metadata))
 }
 
-/// Allocate pages.
-///
-/// The allocated pages are not guarenteed to be contiguous.
-/// The total length of the allocated pages is `len`.
+/// ## Bootstrap page allocator
```
Comment (junyang-zh) [inline ostd/src/mm/page/allocator.rs]: The title should be a plain line without the heading format.

```rust
/// The bootstrapping phase page allocator.
```

File: kernel/libs/page_allocator/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            // Add global free pages to the frame allocator.
+            allocator.add_frame(start, end);
+            info!(
+                "Found usable region, start:{:x}, end:{:x}",
+                region.base(),
+                region.base() + region.len()
+            );
+
+            for frame in start..end {
+                if page::Page::<page::meta::FrameMeta>::check_page_status(frame * PAGE_SIZE) {
```
[the remark this answers is not in the document]
Comment (Tangzh33) [reply kernel/libs/page_allocator/src/lib.rs:41]: We will import `FreePage` handle in the next pr.

File: kernel/libs/page_allocator/src/buddy_allocator.rs [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+    ///
+    /// The function panics if no suitable block found for the given range.
+    ///
+    /// # TODO
+    ///
+    /// We are working on a more efficient implementation, based on free meta
+    /// (i.e., current unused meta) and in the format of linked list. By
+    /// introducing the free meta, we can reduce the time complexity of this
+    /// function from O(log(n)) to O(n).
+    pub fn alloc_specific(&mut self, start: usize, end: usize) {
```
Comment (junyang-zh) [inline kernel/libs/page_allocator/src/buddy_allocator.rs:156]: Do not write `TODO` in the documentation.

Besides, when the linked list is introduced we can only do it `O(n)`, `O(n)` to `O(n)` is not a reduction.

File: osdk/src/base_crate/Cargo.toml.template [hunk 15]
```diff
@@ -2,3 +2,7 @@
 name = "#NAME#"
 version = "#VERSION#"
 edition = "2021"
+
+[dependencies]
+ostd = { version = "0.8.0", path = "../../../ostd" }
```
Comment (junyang-zh) [inline osdk/src/base_crate/Cargo.toml.template:7]: Don't hard code the dependency here. Add them by OSDK. The versions need to match the OSDK version.

File: osdk/test-kernel/Cargo.toml [hunk 16]
```diff
@@ -9,6 +9,6 @@ repository ="https://github.com/asterinas/asterinas"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-ostd = { version = "0.8.1", path = "../../ostd" }
+ostd = { version = "0.8.0", path = "../../ostd" }
```
Comment (junyang-zh) [inline osdk/test-kernel/Cargo.toml:12]: ?

File: ostd/libs/ostd-macros/src/lib.rs [hunk 17]
```diff
@@ -217,3 +217,59 @@ pub fn ktest(_attr: TokenStream, item: TokenStream) -> TokenStream {
...
+/// This macro is used to mark the page allocator initialization function.
+///
+/// # Example
+///
+/// ```ignore
+/// #[ostd::page_allocator_init]
+/// pub fn init_page_allocator() {
+///
+/// }
+/// ```
```
Comment (junyang-zh) [inline ostd/libs/ostd-macros/src/lib.rs:230]: Return anything, even an empty implementation in this function. The users can't tell what to do by reading it.

File: ostd/src/mm/page/mod.rs [hunk 18]
```diff
@@ -83,6 +83,44 @@ impl<M: PageMeta> Page<M> {
...
+    /// # Panics
+    ///
+    /// The function panics if:
+    ///
+    /// - the physical address is out of bound or not aligned.
+    ///
+    /// # TODO
+    ///
+    /// Use a unique handle `FreePage` to check the page status.
+    /// Current implementation is not safe if checking page status and changing page status happens concurrently.
```
Comment (junyang-zh) [inline ostd/src/mm/page/mod.rs:108]: Again, don't write `TODO` in the documentation.

File: kernel/libs/page_allocator/Cargo.toml [hunk 19]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "page_allocator"
```
Comment (junyang-zh) [inline kernel/libs/page_allocator/Cargo.toml:2]: The crate name need to be revised, since we need to publish it and we don't want the user to think it a generic allocator. Name it `aster-page-allocator` instead.

Also, you need to add automatic publishing (`tools/github_workflows/publish_ostd_and_osdk.sh`) and bumping (`tools/bump_version.sh`) for this crate.

File: osdk/src/base_crate/Cargo.toml.template (hunk 15 above)
Comment (junyang-zh) [reply osdk/src/base_crate/Cargo.toml.template:7]: Use `env!(CARGO_PACKAGE_VERSION)` when generating the dependency. You can find similar implementations in OSDK.

File: kernel/libs/aster_page_allocator/src/buddy_allocator.rs [hunk 20]
```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: MPL-2.0
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/buddy_allocator.rs:1]: As @tatetian points out in the comments of #1317 , do not include license or source origin information as Rust doc. Just make them regular comments. Second, it does not say the original code is licensed under MIT. Third, it is unclear which parts are changed by us, which is protected by our copyright license (MPL).

So change the file preamble to the following form.

```rust
// The original source code is from [buddy_system_allocator](https://github.com/rcore-os/buddy_system_allocator),
// which licensed under the following license.
//
// SPDX-License-Identifier: MIT
//
// Copyright (c) 2019-2020 Jiajie Chen
//
// We make the following new changes:
// * Change 1
// * Change 2
//
// These changes are released under the following license:
//
// SPDX-License-Identifier: MPL-2.0
```

File: kernel/libs/aster_page_allocator/src/buddy_allocator.rs [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+
+use alloc::collections::btree_set::BTreeSet;
+use core::{alloc::Layout, array, cmp::min};
+
+use ostd::mm::{page::allocator::PageAlloc, Paddr, PAGE_SIZE};
+
+/// Buddy Frame allocator is a frame allocator based on buddy system, which
+/// allocates memory in power-of-two sizes. The implementation of this allocator
+/// originated from crate [buddy_system_allocator](https://crates.io/crates/buddy_system_allocator).
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/buddy_allocator.rs:25]: Remove the link to the source in the documentation as it is specified in the license preamble.

File: kernel/libs/aster_page_allocator/src/buddy_allocator.rs [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+/// `ORDER`. The frame allocator will only be able to allocate ranges of size
+/// up to 2<sup>ORDER</sup>, out of a total range of size at most 2<sup>ORDER +
+/// 1</sup> - 1.
+pub struct BuddyFrameAllocator<const ORDER: usize = 32> {
+    // buddy system with max order of ORDER
+    free_list: [BTreeSet<usize>; ORDER],
+
+    // statistics
+    pub allocated: usize,
+    pub total: usize,
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/buddy_allocator.rs:39]: These two fields should not be public.

File: kernel/libs/aster_page_allocator/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: MPL-2.0
+
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/lib.rs:2]: Add documentations to the crate to explain what it is.

File: ostd/src/mm/page/mod.rs [hunk 24]
```diff
@@ -83,6 +83,45 @@ impl<M: PageMeta> Page<M> {
         }
     }
 
+    /// Check whether the page described by paddr is allocated.
```
Comment (junyang-zh) [inline ostd/src/mm/page/mod.rs:86]: Use plain text in comments: "Check whether the page at the physical address is allocated".

File: ostd/src/mm/page/mod.rs [hunk 25]
```diff
@@ -83,6 +83,45 @@ impl<M: PageMeta> Page<M> {
...
+    /// true if the page is allocated, false otherwise.
+    ///
+    /// # Safety
+    ///
+    /// Page status checked by this function may be outdated, since other running threads may change them after the atomic load.
```
Comment (junyang-zh) [inline ostd/src/mm/page/mod.rs:91]: Better to adhere to the 80 characters width convention here.

File: ostd/src/mm/page/mod.rs [hunk 26]
```diff
@@ -83,6 +83,45 @@ impl<M: PageMeta> Page<M> {
...
+    /// The function panics if:
+    ///
+    /// - the physical address is out of bound or not aligned.
+    ///
+    // # TODO
+    //
+    // Use a unique handle `FreePage` to check the page status.
+    // Current implementation is not safe if checking page status and changing
+    // page status happens concurrently.
+    pub fn check_page_status(paddr: Paddr) -> bool {
```
Comment (junyang-zh) [inline ostd/src/mm/page/mod.rs:110]: Rename it to `is_page_allocated`.

File: tools/bump_version.sh [hunk 27]
```diff
@@ -130,6 +132,8 @@ update_dep_version ${OSTD_CARGO_TOML_PATH} ostd-macros
 update_package_version ${OSDK_CARGO_TOML_PATH}
 update_package_version ${OSDK_TEST_RUNNER_CARGO_TOML_PATH}
 update_dep_version ${OSDK_TEST_RUNNER_CARGO_TOML_PATH} ostd
+update_package_version ${ASTER_KERNEL_PATH}/libs/aster_page_allocator/Cargo.toml
+update_dep_version ${ASTER_KERNEL_TOML_PATH} aster_page_allocator
```
Comment (junyang-zh) [inline tools/bump_version.sh:136]: The kernel don't specify a version for it. You can remove this line.

File: ostd/src/mm/page/allocator.rs [hunk 28]
```diff
@@ -5,70 +5,86 @@
...
-            allocated: 0,
-        }
-    }
+    /// Allocates a contiguous range of pages described by the layout.
+    ///
+    /// # Panics
+    ///
+    /// The function panics if the layout.size is not base-page-aligned or
+    /// if the layout.align is less than the PAGE_SIZE.
+    fn alloc(&mut self, layout: Layout) -> Option<Paddr>;
```
Comment (junyang-zh) [inline ostd/src/mm/page/allocator.rs:45]: We've just decided that the page allocator trait should enable lock-free design of local page allocation cache. So the signature should be `fn alloc(&self, layout: Layout) -> Option<Paddr>;`. Similarly, the `alloc_page` and `dealloc` functions takes shared references. The implementation should manage it's own locking scheme. For now, you can make a inner big lock around the global free list, and make the statistics atomic counters.

In the future, we can maintain per-cpu free lists to allow lock-free design.

Comment (Tangzh33) [reply ostd/src/mm/page/allocator.rs:45]: Get it. I will refactor the corresponding trait in the subsequent pr. Next pr will furtherly enhance the performance of page allocator, as well as refactor the corresponding trait to meet the latest design.

File: kernel/libs/aster_page_allocator/src/lib.rs [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+//! This crate provides a global page allocator for the kernel.
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/lib.rs:3]: //! An implementation of the global page allocator for OSTD based kernels.

File: kernel/libs/aster_page_allocator/src/lib.rs [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+//!
+//! Currently, the page allocator is implemented by the buddy system.
+//! By corporating with the `PageAlloc` trait and `page_allocator_init`
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/lib.rs:6]: Still it is important to polish documentations in the published crates. You can refer to the OSTD API in the documentation by such syntax:

```rust
//! ...
//! By corporating with the [`PageAlloc`] trait and the [`page_allocator_init`]
//! procedural macro.
//! ...
//!
//! [`PageAlloc`]: ostd::mm::PageAlloc
//! [`page_allocator_init`]: ostd::page_allocator_init
```

File: kernel/libs/aster_page_allocator/src/lib.rs [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+//! procedure macro, page allocator's implementation can be decoupled from the
+//! ostd and can be easily replaced to serve desinated purposes.
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/lib.rs:8]: `[OSTD](https://crates.io/crates/ostd)`

And there's a typo "desinated". Please double check if the typo checker can cover this crate.

File: tools/bump_version.sh [hunk 32]
```diff
@@ -109,6 +109,8 @@ OSTD_MACROS_CARGO_TOML_PATH=${ASTER_SRC_DIR}/ostd/libs/ostd-macros/Cargo.toml
 OSDK_CARGO_TOML_PATH=${ASTER_SRC_DIR}/osdk/Cargo.toml
 OSDK_TEST_RUNNER_CARGO_TOML_PATH=${ASTER_SRC_DIR}/osdk/test-kernel/Cargo.toml
 VERSION_PATH=${ASTER_SRC_DIR}/VERSION
+ASTER_KERNEL_PATH=${ASTER_SRC_DIR}/kernel
+ASTER_KERNEL_TOML_PATH=${ASTER_KERNEL_PATH}/Cargo.toml
```
Comment (junyang-zh) [inline tools/bump_version.sh:113]: This variable is not referenced anymore, remove it.

File: kernel/libs/aster_page_allocator/src/buddy_allocator.rs [hunk 33]
```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier: MPL-2.0 OR MIT
+// To be specific, the original source code is from [buddy_system_allocator](https://github.com/rcore-os/buddy_system_allocator),
```
Comment (junyang-zh) [inline kernel/libs/aster_page_allocator/src/buddy_allocator.rs:2]: Insert a blank line between 1 and 2. Also break the line before the link to improve readability.

File: kernel/libs/aster_page_allocator/src/lib.rs (hunk 31 above)
Comment (Tangzh33) [reply kernel/libs/aster_page_allocator/src/lib.rs:8]: Typo checker can cover this crate, but the typo "desinated" cannot be detected. Probably because the typo checker's dictionary is imperfect.

**Review by junyang-zh — APPROVED** [review]
Looks good to me at this stage. Let's invite @tatetian for further reviews.

**Comment by junyang-zh** [comment]
After the approval from @tatetian , please bump the version by a minor version number with a commit.

File: ostd/src/mm/page/meta.rs [hunk 34]
```diff
@@ -306,18 +311,15 @@ pub(crate) fn init() -> Vec<Page<MetaPageMeta>> {
...
-        .lock()
-        .alloc(core::alloc::Layout::from_size_align(nframes * PAGE_SIZE, PAGE_SIZE).unwrap())
-        .expect("Failed to allocate metadata pages");
-    // Zero them out as initialization.
-    let vaddr = paddr_to_vaddr(start_frame) as *mut u8;
-    unsafe { core::ptr::write_bytes(vaddr, 0, PAGE_SIZE * nframes) };
-    for i in 0..nframes {
-        let paddr = start_frame + i * PAGE_SIZE;
-        meta_pages.push(paddr);
+    let mut allocator = BOOTSTRAP_PAGE_ALLOCATOR.get().unwrap().lock();
```
Comment (tatetian) [inline ostd/src/mm/page/meta.rs:319]: Here, we are using `BOOTSTRAP_PAGE_ALLOCATOR` explicitly.

```rust
    let mut meta_pages = Vec::new();
```

But in the first line of this function, we are doing heap allocation by using `Vec`. Where does the pages used by heap come from? Do we have any mechanism that prevents the heap allocator from using `PAGE_ALLOCATOR`?

**Review by tatetian — COMMENTED** [review]
Good job on pulling this huge refactoring work! The PR is in a good shape. I have raised some minor issues that I would like to discuss with you. Thanks!

File: ostd/src/mm/page/allocator.rs [hunk 35]
```diff
@@ -80,53 +89,179 @@ pub(crate) fn alloc_single<M: PageMeta>(metadata: M) -> Option<Page<M>> {
...
+/// `ORDER`. The frame allocator will only be able to allocate ranges of size
+/// up to 2<sup>ORDER</sup>, out of a total range of size at most 2<sup>ORDER +
+/// 1</sup> - 1.
+pub struct BuddyFrameAllocator<const ORDER: usize = 32> {
+    // buddy system with max order of ORDER
+    free_list: [BTreeSet<usize>; ORDER],
+
+    // statistics
+    pub allocated: usize,
+    pub total: usize,
```
[the remark this answers is not in the document]
Comment (Tangzh33) [reply ostd/src/mm/page/allocator.rs]: Get it. Originally these are temporarily debug interfaces. I have already fix it in the following commits but forgot to clear the history. I will make it right.

File: ostd/src/mm/page/allocator.rs [hunk 36]
```diff
@@ -5,70 +5,79 @@
...
     }
 
-    pub fn dealloc(&mut self, start_frame: usize, count: usize) {
-        self.allocator.dealloc(start_frame, count);
-        self.allocated -= count * PAGE_SIZE;
-    }
+    /// Deallocates a specified number of consecutive pages.
+    ///
+    /// Warning! May lead to panic when afterwards allocation while using
+    /// out-of `ostd`
```
[the remark this answers is not in the document]
Comment (Tangzh33) [reply ostd/src/mm/page/allocator.rs:68]: Get it. I have rewritten the sentence as follows:

> /// # Warning
    ///
    /// In `ostd`, the correctness of the allocation / deallocation is
    /// guaranteed by the meta system ( [`crate::mm::page::meta`] ), while the
    /// page allocator is only responsible for managing the allocation
    /// metadata. The typed system can only be used within the `ostd` crate.
    ///
    /// Therefore, deallocating pages out-of `ostd` without coordination with
    /// the meta system may lead to unexpected behavior, such as panics during
    /// afterwards allocation.

File: osdk/src/base_crate/mod.rs [hunk 37]
```diff
@@ -89,6 +89,14 @@ pub fn new_base_crate(
     // Add dependencies to the Cargo.toml
     add_manifest_dependency(dep_crate_name, dep_crate_path, link_unit_test_runner);
 
+    // Set the dependencies of base crate
+    let manifest_path = PathBuf::from(env!("CARGO_MANIFEST_DIR").to_string());
+    add_base_dependency("ostd", manifest_path.join("../ostd"));
+    add_base_dependency(
+        "aster_page_allocator",
+        manifest_path.join("../kernel/libs/aster_page_allocator"),
```
[the remark this answers is not in the document]
Comment (Tangzh33) [reply osdk/src/base_crate/mod.rs:97]: Sorry for the confusion. As you have pointed out, `page_allocator` is indeed a typo. I have corrected it in the latest version.

File: ostd/src/mm/page/allocator.rs [hunk 38]
```diff
@@ -234,6 +233,74 @@ impl<const ORDER: usize> BuddyFrameAllocator<ORDER> {
...
+    /// # Description
+    ///
+    /// Given frames, described by a range of **frame number** [start, end),
+    /// mark them as allocated. Make sure they can be correctly deallocated
+    /// afterwards, while will not be allocated before deallocation.
+    ///
+    /// # Panics
+    ///
+    /// The function panics if no suitable block found for the given range.
+    pub fn set_frames_allocated(&mut self, start: usize, end: usize) {
```
[the remark this answers is not in the document]
Comment (Tangzh33) [reply ostd/src/mm/page/allocator.rs]: It is fine if the caller violates the pre-condition above since the function has already considered this. 

Essentially, the function traverses all the free lists until it finds one free block containing at least one input-described frame. If there is no suitable block, the function will panic. 

It is important to note that the free lists can only be initialized through the function `add_frame`. Therefore, all the found blocks come from the frames previously added by calling `add_frame`. You can refer to the code and the comments within for more details : )

File: ostd/src/mm/page/allocator.rs [hunk 39]
```diff
@@ -282,3 +349,121 @@ pub(crate) fn init() {
...
+    }
+
+    fn alloc_page(&mut self, _align: usize) -> Option<Paddr> {
+        let frame = self.frame_cursor;
+        debug!("allocating frame: {:#x}", frame * PAGE_SIZE,);
+        // Update idx and cursor
+        let regions = crate::boot::memory_regions();
+        self.frame_cursor += 1;
+        loop {
+            let region = regions[self.mem_region_idx];
```
[the remark this answers is not in the document]
Comment (Tangzh33) [reply ostd/src/mm/page/allocator.rs:212]: According to the definition, the memory region is a global static vector `Vec<MemoryRegion>` while `boot::memory_regions()` is its getter function, returning the vector's reference. 

If we need to cache `region`s' information, it is natural to copy it to a local vector, which is similar to how `region`s' information is originally stored and used. Therefore, whether cache the `region` in the bootstrap allocator's context or not will have a subtle impact on the function's performance. 

Besides, since the heap space is limited while bootstrapping, we should avoid using heap space in the bootstrap allocator, the minimal implementation of page allocator.

File: ostd/src/mm/page/meta.rs (hunk 34 above)
Comment (Tangzh33) [reply ostd/src/mm/page/meta.rs:319]: According to the previous implementation, the initial space managed by the heap allocator comes from a const vector sized 256 pages:

``` rust
const INIT_KERNEL_HEAP_SIZE: usize = PAGE_SIZE * 256;

#[repr(align(4096))]
struct InitHeapSpace([u8; INIT_KERNEL_HEAP_SIZE]);

static mut HEAP_SPACE: InitHeapSpace = InitHeapSpace([0; INIT_KERNEL_HEAP_SIZE]);
```

The heap allocator will function correctly before the initial space is burnt out. If all the initial space is used, the heap allocator will call its `rescue` function to apply new pages from the page allocator. 

The kernel may crash if `PAGE_ALLOCATOR` is not ready while the heap needs new pages. Since only `meta` will use heap memory before `PAGE_ALLOCTOR` is initialized, while the memory usage of meta can be estimated, it is of small possibility that the kernel will crash due to the aforementioned reasons.

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by Tangzh33** [comment]
@tatetian Sorry for the delay! I have been working hard to follow the modification merged before, which removes the unstable `linkage` feature. The `linkage` feature, which provides both strong and weak linkage to the same symbols, is imported to make the regular injected page allocator compatible with the test kernel's injected allocator.

Without the feature, we require OSTD's users to explicitly insert the procedure macro annotations to implement their own injected page allocator or provide a default page allocator, often functional but inefficient, for users to decide whether to inject. After discussing it with @junyang-zh, the previous solution is far from elegant, while the latter necessitates supporting the following commits in PR #1418.

So, all of the effort on the latter choice must be reverted temporarily due to the failure of OSDK's CI. Luckily, only some of the recent codes have been deprecated. I plan to improve the injection method continuously in the following PR. The following PR will inject the page allocator with functions instead of symbols and macros, and the kernel will function correctly without explicit injection.

Sorry again for the unintended delay : ( The PR is currently ready to be merged. One of the CI failures is due to the new procedure macro, and the bump version will fix it, while the other randomly fails and cannot be reproduced on my machine.

**Comment by Tangzh33** [comment]
@boterinas rerun

**Comment by junyang-zh** [comment]
Prefer #1783
